### PR TITLE
Add message saying whether username is available while user is typing

### PR DESCRIPTION
**Pull request by ldcWV**

### CODE REVIEW

File: src/main/webapp/js/profilePage.js [hunk 1]
```diff
@@ -1,13 +1,79 @@
...
+    if(event.keyCode === 13) { // enter button
+      document.getElementById('submit-username-button').click();
+    } else {
+      clearTimeout(timer);
+      timer = setTimeout(displayAvailability, waitTime);
+    }
+  });
+}
+
+async function isTaken(desiredUsername) {
```
Comment (saramissak) [inline src/main/webapp/js/profilePage.js:33]: Add JS doc for this.

File: src/main/webapp/js/profilePage.js [hunk 2]
```diff
@@ -1,13 +1,79 @@
...
+  params.append('username', desiredUsername);
+
+  let result;
+  await fetch('/username-available-data', {method: 'post', body: params}).then(response => response.json()).then(taken => {
+    result = taken;
+  });
+  return result;
+}
+
+async function getAvailabilityText(desiredUsername) {
```
Comment (saramissak) [inline src/main/webapp/js/profilePage.js:49]: Add JS doc

File: src/main/webapp/js/profilePage.js [hunk 3]
```diff
@@ -1,13 +1,79 @@
...
+  if(desiredUsername.length == 0) {
+    return 'Username must be at least 1 character'
+  }
+  if(desiredUsername.length > 20) {
```
Comment (saramissak) [inline src/main/webapp/js/profilePage.js:53]: This is okay but any reason why 20?

File: src/main/webapp/js/profilePage.js [hunk 4]
```diff
@@ -1,13 +1,79 @@
...
+  if(!pattern.test(desiredUsername)) {
+    return 'Only letters, digits, underscores, and periods are allowed';
+  }
+  if(await isTaken(desiredUsername)) {
+    return 'Username is already taken'
+  }
+  return 'Available';
+}
+
+async function displayAvailability() {
```
Comment (saramissak) [inline src/main/webapp/js/profilePage.js:69]: JS Doc please.

File: src/main/webapp/js/profilePage.js [hunk 5]
```diff
@@ -1,13 +1,79 @@
...
+  availabilityBox.innerText = await getAvailabilityText(desiredUsername);
+  if(availabilityBox.innerText == 'Available') {
+      availabilityBox.className = 'green-text';
+  } else {
+      availabilityBox.className = 'red-text';
+  }
+}
+
+async function submitUsername() {
+  desiredUsername = document.getElementById('userName').value;
```
Comment (saramissak) [inline src/main/webapp/js/profilePage.js]: let*

File: src/main/webapp/js/profilePage.js [hunk 6]
```diff
@@ -1,13 +1,79 @@
...
+  let availabilityBox = document.getElementById('username-availability-message');
+  availabilityBox.innerText = await getAvailabilityText(desiredUsername);
+  if(availabilityBox.innerText == 'Available') {
+    const params = new URLSearchParams();
+    params.append('userName', desiredUsername);
+    await fetch('/create-username-data', {method: 'post', body: params});
+    window.location = '/profilePage.html'
+  } else {
+    alert(availabilityBox.innerText);
```
Comment (saramissak) [inline src/main/webapp/js/profilePage.js:90]: Does this have to be an alert, is there a better way to do this for UX?

File: src/main/webapp/js/profilePage.js [hunk 7]
```diff
@@ -1,13 +1,79 @@
...
+  }
 }
```
Comment (saramissak) [inline src/main/webapp/js/profilePage.js]: Add line at end of file please.

File: src/main/webapp/js/profilePage.js [hunk 8]
```diff
@@ -1,13 +1,79 @@
...
+  }
+  if(desiredUsername.length > 20) {
+    return 'Username cannot be more than 20 characters'
+  }
+  let pattern = new RegExp(/^[ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz._0123456789]*$/);
+  if(!pattern.test(desiredUsername)) {
+    return 'Only letters, digits, underscores, and periods are allowed';
+  }
+  if(await isTaken(desiredUsername)) {
+    return 'Username is already taken'
```
Comment (saramissak) [inline src/main/webapp/js/profilePage.js]: Add ";" to the end

File: src/main/webapp/js/profilePage.js [hunk 9]
```diff
@@ -1,13 +1,79 @@
...
+  }
+  return 'Available';
```
Comment (saramissak) [inline src/main/webapp/js/profilePage.js:63]: Add ";" to the end

File: src/main/webapp/js/profilePage.js [hunk 10]
```diff
@@ -1,13 +1,79 @@
...
+  let result;
+  await fetch('/username-available-data', {method: 'post', body: params}).then(response => response.json()).then(taken => {
+    result = taken;
+  });
+  return result;
+}
+
+async function getAvailabilityText(desiredUsername) {
+  if(desiredUsername.length == 0) {
+    return 'Username must be at least 1 character'
```
Comment (saramissak) [inline src/main/webapp/js/profilePage.js]: Add ";" to the end

File: src/main/webapp/profilePage.html [hunk 11]
```diff
@@ -33,16 +33,16 @@
...
-            <h5>Username</h5>
-            <input id='userName' type='text' name='userName' required>
-          </div>
-          <br>
-          <input type="submit" value="Submit">
-        </form>        
+      <h1>Profile</h1>
     </div>
+    <div class='input-field'>
+    <h5>Username</h5>
```
Comment (saramissak) [inline src/main/webapp/profilePage.html]: indent by 2 more spaces here and line 40

File: src/main/webapp/js/profilePage.js [hunk 12]
```diff
@@ -1,13 +1,79 @@
...
+
+async function submitUsername() {
+  desiredUsername = document.getElementById('userName').value;
+  let availabilityBox = document.getElementById('username-availability-message');
+  availabilityBox.innerText = await getAvailabilityText(desiredUsername);
+  if(availabilityBox.innerText == 'Available') {
+    const params = new URLSearchParams();
+    params.append('userName', desiredUsername);
+    await fetch('/create-username-data', {method: 'post', body: params});
+    window.location = '/profilePage.html'
```
Comment (saramissak) [inline src/main/webapp/js/profilePage.js]: ";" at the end

File: src/main/webapp/js/profilePage.js (hunk 1 above)
Comment (ldcWV) [reply src/main/webapp/js/profilePage.js:33]: Done, thanks.

File: src/main/webapp/js/profilePage.js (hunk 2 above)
Comment (ldcWV) [reply src/main/webapp/js/profilePage.js:49]: Done, thanks.

File: src/main/webapp/js/profilePage.js (hunk 3 above)
Comment (ldcWV) [reply src/main/webapp/js/profilePage.js:53]: We probably need a cap on the length; there could be problems with displaying the username if it's too long. As for the number 20, it's chosen arbitrarily. Seems like a good cap.

File: src/main/webapp/js/profilePage.js (hunk 4 above)
Comment (ldcWV) [reply src/main/webapp/js/profilePage.js:69]: Done, thanks.

File: src/main/webapp/js/profilePage.js (hunk 5 above)
Comment (ldcWV) [reply src/main/webapp/js/profilePage.js]: Done, thanks.

File: src/main/webapp/js/profilePage.js (hunk 6 above)
Comment (ldcWV) [reply src/main/webapp/js/profilePage.js:90]: I think an alert for this looks pretty decent. If you have a better idea please let me know.

File: src/main/webapp/js/profilePage.js (hunk 7 above)
Comment (ldcWV) [reply src/main/webapp/js/profilePage.js]: Done, thanks.

File: src/main/webapp/js/profilePage.js (hunk 8 above)
Comment (ldcWV) [reply src/main/webapp/js/profilePage.js]: Done, thanks.

File: src/main/webapp/js/profilePage.js (hunk 9 above)
Comment (ldcWV) [reply src/main/webapp/js/profilePage.js:63]: Done, thanks.

File: src/main/webapp/js/profilePage.js (hunk 10 above)
Comment (ldcWV) [reply src/main/webapp/js/profilePage.js]: Done, thanks.

File: src/main/webapp/profilePage.html (hunk 11 above)
Comment (ldcWV) [reply src/main/webapp/profilePage.html]: Done, thanks.

File: src/main/webapp/js/profilePage.js (hunk 12 above)
Comment (ldcWV) [reply src/main/webapp/js/profilePage.js]: Done, thanks.

File: src/main/webapp/js/profilePage.js (hunk 3 above)
Comment (saramissak) [reply src/main/webapp/js/profilePage.js:53]: I agree, this seems good.

File: src/main/webapp/js/profilePage.js (hunk 6 above)
Comment (saramissak) [reply src/main/webapp/js/profilePage.js:90]: Maybe show it to me tomorrow over GVC? Also, I noticed that when you delete characters it doesn't recheck if it's good or not, is that something we can fix? So for instance, if I type saraM it says available but if I put a backspace to Sara it doesn't say its unavailable even though it isn't available

**Review by saramissak — APPROVED** [review]

**Pull request merged**